lib/http/node.js

```diff
@@ -76,15 +75,40 @@ AWS.NodeHttpClient = AWS.util.inherit({
 
   writeBody: function writeBody(stream, httpRequest) {
     var body = httpRequest.body;
-
-    if (body && WritableStream && ReadableStream) { // progress support
-      if (!(body instanceof Stream)) body = AWS.util.buffer.toStream(body);
-      body.pipe(this.progressStream(stream, httpRequest));
-    }
+    var totalBytes = parseInt(httpRequest.headers['Content-Length'], 10);
+    var loadedBytes = 0;
 
     if (body instanceof Stream) {
+      // for progress support of streaming content
+      // tap the data event of the stream in addition to piping
+      body.on('data', function(chunk) {
+        if (chunk) {
+          loadedBytes += chunk.length;
+          stream.emit('sendProgress', {
+            loaded: loadedBytes,
+            total: totalBytes
+          });
+        }
+      });
       body.pipe(stream);
     } else if (body) {
+      // the provided body is a buffer/string and is already fully available in memory,
+      // so for performance it's best to send it as a whole by calling stream.end(body),
+      // but since callers expect some progression, we emit one progress event before sending,
+      // and another once the stream drains.
+      // the use of totalBytes is important over body.length for strings where length is char length and not byte length.
+      loadedBytes = totalBytes;
+      stream.emit('sendProgress', {
+        loaded: loadedBytes,
+        total: totalBytes
+      });
+      stream.once('drain', function() {
```

Comment:
I don't think it's important to emit `sendProgress` twice. Our docs simply state that the `httpUploadProgress` event will be triggered when data is uploaded. Emitting once `drain` is called is probably more accurate, since that indicates that the data has actually been uploaded.

I suspect that buffers/strings were converted to a stream so that the user had more fine-grained info on how much data was sent. However, I doubt it's worth the performance trade-off.